# SHORTEST NEURAL NETWORK CODE. (FOR BEGINNERS)

All you need is one library 'tensorflow'. Ofcourse we are using Keras(Highlevel API). It is now officially included in tesorflow

In [77]:
import tensorflow as tf

Load the MNIST dataset. The data has train and test splits with the features and labels separated.

In [78]:
(tr_x,tr_y),(ts_x,ts_y) = tf.keras.datasets.mnist.load_data()

In [79]:
print(tr_x.shape , tr_y.shape , ts_x.shape, ts_y.shape)

(60000, 28, 28) (60000,) (10000, 28, 28) (10000,)


Examine the first image in train set.

The input values are in range O-255(0 being pure white, 255 dark black). 
So scale them to 0-1 for faster calculation. (Advised, not mandatory).

Execute Print(tr_x[0]) to know about this.

In [80]:
tr_x = tr_x/255
ts_x = ts_x/255

Since, NN's need data to be flattened. Reshape the data.

In [81]:
import numpy as np
flat_tr_x = np.reshape(tr_x,[60000,784])
flat_ts_x = np.reshape(ts_x,[10000,784])

In [82]:
print(flat_tr_x.shape, flat_ts_x.shape)

(60000, 784) (10000, 784)


NEURAL NETWORK PART: (A feed forward NN)

In [83]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(140,activation = 'relu'))
model.add(tf.keras.layers.Dense(220,activation = 'relu'))
model.add(tf.keras.layers.Dense(10,activation = 'softmax'))

Select your loss function and the type of optimizer to minimize it.(check keras website various losses and optimizers)

In [84]:
model.compile(optimizer = 'Adagrad',loss = 'sparse_categorical_crossentropy', metrics =['accuracy'])

Fit the flattened data.

In [85]:
model.fit(flat_tr_x,tr_y,epochs = 5,verbose = 1)

Epoch 1/5
60000/60000 [==============================] - 10s 171us/step - loss: 0.2033 - acc: 0.9394
Epoch 2/5
60000/60000 [==============================] - 10s 159us/step - loss: 0.1015 - acc: 0.9698
Epoch 3/5
60000/60000 [==============================] - 9s 156us/step - loss: 0.0768 - acc: 0.9772
Epoch 4/5
60000/60000 [==============================] - 10s 159us/step - loss: 0.0625 - acc: 0.9818
Epoch 5/5
60000/60000 [==============================] - 10s 166us/step - loss: 0.0529 - acc: 0.9846


You could see, loss is decreasing gradually and the accuracy is increasing.

In [86]:
print(model.evaluate(flat_ts_x,ts_y))

10000/10000 [==============================] - 1s 96us/step
[0.07530616769352928, 0.9748]


The Testing accuracy is found to be 0.9748. Dont be happy, it is pretty common with MNIST data.
CNN can achieve about 99% accuracy. 
Check my code on CNN.

The tensorflow tutorial of the same process is also available in my github check it.

Inspired from Sentdex.

# This is shortest NN code (Quite basic code) for beginners.
